### PR TITLE
[Ability] Partially implement Gulp Missile

### PR DESCRIPTION
**Pull request by torranx**
## What are the changes?
partially implements gulp missile:
**Does not transform when Surf/Dive misses/is protected**

## Why am I doing these changes?
unimplemented ability

## What did change?
- added new abattr class `PostDefendGulpMissileAbAttr` 
- added new battlertag `GulpMissileTag`
- added new move attr `GulpMissileTagAttr`

### Screenshots/Videos
- Gulping form

https://github.com/user-attachments/assets/2723ffe4-1770-413f-a611-7a7e2f5e1f03

- Gorging form

https://github.com/user-attachments/assets/a3821217-24ea-426e-a763-aa24f509a0ec

## How to test the changes?
overrides used: 
```
{
  BATTLE_TYPE_OVERRIDE: "single",
  STARTER_SPECIES_OVERRIDE: Species.CRAMORANT,
  ABILITY_OVERRIDE: Abilities.GULP_MISSILE,
  MOVESET_OVERRIDE: [Moves.DIVE, Moves.SURF, Moves.SPLASH, Moves.QUICK_ATTACK],
  OPP_LEVEL_OVERRIDE: 5,
  OPP_MOVESET_OVERRIDE: Array(4).fill(Moves.TACKLE)
}
```
`$ npm run test gulp_missile`

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

**Comment by torranx** [comment]
Moving this to draft because I mistakenly removed the test involving Magic Guard. Will possibly update Dive and Surf score calculation too.

File: src/data/ability.ts [hunk 1]
```diff
@@ -496,6 +496,49 @@ export class PostDefendAbAttr extends AbAttr {
...
+    const battlerTag = pokemon.getTag(GulpMissileTag);
+    if (!battlerTag || move.category === MoveCategory.STATUS) {
+      return false;
+    }
+
+    const cancelled = new Utils.BooleanHolder(false);
+    applyAbAttrs(BlockNonDirectDamageAbAttr, attacker, cancelled);
+
+    if (!cancelled.value) {
+      attacker.damageAndUpdate(Math.floor(attacker.getMaxHp() * 1/4), HitResult.OTHER);
```
Comment (Tempo-anon) [inline src/data/ability.ts:528]: nit: I think there should also be a `Math.max(Math.floor(attacker.getMaxHp() / 4), 1)` so Shedinja doesn't just ignore it

File: src/data/battler-tags.ts [hunk 2]
```diff
@@ -1647,6 +1647,47 @@ export class StockpilingTag extends BattlerTag {
...
+    super(tagType, BattlerTagLapseType.CUSTOM, 0, sourceMove);
+  }
+
+  /**
+   * Gulp Missile's initial form changes are triggered by using Surf and Dive.
+   * @param {Pokemon} pokemon The Pokemon with Gulp Missile ability.
+   * @returns Whether the BattlerTag can be added.
+   */
+  canAdd(pokemon: Pokemon): boolean {
+    const isSurfOrDive= [ Moves.SURF, Moves.DIVE ].includes(this.sourceMove);
```
Comment (Tempo-anon) [inline src/data/battler-tags.ts:1665]: nit: space after `isSurfOrDive`

File: src/data/battler-tags.ts [hunk 3]
```diff
@@ -1647,6 +1647,47 @@ export class StockpilingTag extends BattlerTag {
...
+    return isSurfOrDive && !pokemon.getTag(BattlerTagType.GULP_MISSILE_ARROKUDA) && !pokemon.getTag(BattlerTagType.GULP_MISSILE_PIKACHU);
```
Comment (Tempo-anon) [inline src/data/battler-tags.ts:1666]: This tag can be added on Pokemon that aren't Cramorant right? Will this lead to weird interactions like it being baton passable?

File: src/data/move.ts [hunk 4]
```diff
@@ -4276,6 +4276,45 @@ export class AddBattlerTagAttr extends MoveEffectAttr {
...
+   * @param {Move} move The move being used.
+   * @param {any[]} args Additional arguments, if any.
+   * @returns Whether the BattlerTag is applied.
+   */
+  apply(user: Pokemon, target: Pokemon, move: Move, args: any[]): boolean | Promise<boolean> {
+    if (!super.apply(user, target, move, args)) {
+      return false;
+    }
+
+    if (user.hasAbility(Abilities.GULP_MISSILE)) {
```
Comment (Tempo-anon) [inline src/data/move.ts:4301]: Should there also be the species check for Cramorant here? Or is that a really unnecessary optimization?

File: src/data/battler-tags.ts (hunk 3 above)
Comment (torranx) [reply src/data/battler-tags.ts]: added species check

File: src/data/move.ts (hunk 4 above)
Comment (torranx) [reply src/data/move.ts]: added species check

File: src/test/abilities/gulp_missile.test.ts [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    game.override
+      .battleType("single")
+      .moveset([Moves.SURF, Moves.DIVE, Moves.SPLASH])
+      .enemySpecies(Species.SNORLAX)
+      .enemyAbility(Abilities.BALL_FETCH)
+      .enemyMoveset(SPLASH_ONLY)
+      .enemyLevel(5);
+  });
+
+  it("changes to Gulping Form if HP is over half when Surf or Dive is used", async () => {
```
Comment (innerthunder) [inline src/test/abilities/gulp_missile.test.ts:46]: According to [Smogon](https://www.smogon.com/dex/sv/abilities/gulp-missile/), Gulp Missile's effects trigger on the first turn of Dive.

Comment (torranx) [reply src/test/abilities/gulp_missile.test.ts:46]: I've checked videos as references and it only transforms after "coming out of dive" - which thematically should be expected too. Also, functionality-wise I don't think it matters.

Comment (Tempo-anon) [reply src/test/abilities/gulp_missile.test.ts:46]: I think there are edge cases that will happen. No guard and Surf for example can hit a diving Pokemon.

Comment (torranx) [reply src/test/abilities/gulp_missile.test.ts:46]: Tested the ff scenarios on showdown:

- Cramorant in gulping form and underwater is hit by surf on TURN 2 (gulp missile does not proc)
image


- Cramorant in gulping form and underwater is hit by surf on TURN 1 (gulp missle does not proc)
image

It seems Gulp Missile only activates when not underwater. I will test the above scenarios with my changes and update here.

Comment (torranx) [reply src/test/abilities/gulp_missile.test.ts:59]: Tested the above scenarios and works as intended:
- Cramorant in gulping form and underwater is hit by surf on TURN 2 (gulp missile does not proc)


https://github.com/user-attachments/assets/3095a890-9ae8-4502-8022-69aa3132f34b



- Cramorant in gulping form and underwater is hit by surf on TURN 1 (gulp missle does not proc)


https://github.com/user-attachments/assets/928a9b9c-3b54-417a-83f8-f944f4443cd5



Added automated test of the scenario too

Comment (torranx) [reply src/test/abilities/gulp_missile.test.ts:59]: Dive has 100% accuracy

Comment (innerthunder) [reply src/test/abilities/gulp_missile.test.ts:59]: [Testing Dive into Protect on Showdown](https://github.com/user-attachments/files/16501748/Gen9OU-2024-08-05-inthunder-innerthunder.zip)

Cramorant changes form even if Dive fails to hit the target.

File: src/test/abilities/gulp_missile.test.ts [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+  const GORGING_FORM = 2;
+
+  /**
+   * Gets the effect damage of Gulp Missile
+   * See Gulp Missile {@link https://bulbapedia.bulbagarden.net/wiki/Gulp_Missile_(Ability)}
+   * @param {Pokemon} pokemon The pokemon taking the effect damage.
+   * @returns The effect damage of Gulp Missile
+   */
+  const getEffectDamage = (pokemon: Pokemon): number => {
+    return Math.max(Math.floor(pokemon.getMaxHp() * 1/4));
```
Comment (Tempo-anon) [inline src/test/abilities/gulp_missile.test.ts:35]: Is this missing a `, 1` since it looks like you're just max'ing a single value. A good test for this would be seeing if this KOs a Shedinja. Otherwise this PR looks fine, is there a reason it's still in draft?

Comment (torranx) [reply src/test/abilities/gulp_missile.test.ts:35]: yeah missed this, it's in draft because if the dive missed, it does not change form (tho it works on water absorb). I have a solution for that already but I'm not quite happy with it.

Comment (torranx) [reply src/test/abilities/gulp_missile.test.ts:35]: Yes, workaround is call the Gulp Missile move attr in the else statement of ChargeAttr (but I dont find it pretty and haven't tested it much). ~I could fix it by slapping a (P) on it~
